### ISSUE TITLE
How to implement train_step with multiple gradient calculations with a JAX backend? Ex. GAN

### ISSUE THREAD

**Issue by craig-sony** [keras-team/keras#18881]
I have been trying to figure out how to write a GAN using Keras 3 with a JAX backend using the `stateless_call` API.
I cannot figure out a clean way to deal with the need to have separate gradients computed for the discriminator and generator.
The only approach I've gotten close with is to create a mapping between the trainable_variables/non_trainable_variables lists used by the model and the corresponding layers. Then when I call `stateless_call` I first have to extract the corresponding trainable_variables/non_trainable_variables for the layer being called from those passed into the `train_step` function, but then I need to reinsert the `non_trainable_variables` returned by `stateless_call`. It's a mess.

Can you please update the following example for Keras 3?
https://keras.io/examples/generative/conditional_gan/

Thanks.

**Comment by fchollet** [comment]
For a real-world example see how we handle stateful metrics in the JAX backend: https://github.com/keras-team/keras/blob/master/keras/backend/jax/trainer.py#L130-L145

In general, working with JAX statelessness is pretty terrible, so the solution is to open a StatelessScope and pretend everything is stateful 👍

**Comment by craig-sony** [comment]
Thanks for the info, any chance of getting the example updated though?

**Comment by dhantule** [comment]
Hi @craig-sony, thanks for reporting this. The [ example](https://keras.io/examples/generative/conditional_gan/) has been updated and runs fine with keras 3 in this [gist](https://colab.sandbox.google.com/gist/dhantule/c963907c9a63ffe59ba4691679909706/example-conditional_gan.ipynb).
